### PR TITLE
Update NUOPC/CMEPS driver

### PR DESCRIPTION
**Pull request by DeniseWorthen**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [x] Short (1 sentence) summary of your PR: 
Updates drivers/nuopc/cmeps for both CESM and UFS
- [x] Developer(s): 
    @DeniseWorthen 
    @dabail10 
    @mvertens 
- [x] Suggest PR reviewers from list in the column to the right.
    @dabail10 
    @apcraig 
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Regression test logs for Cheyenne [Intel](https://github.com/DeniseWorthen/ufs-weather-model/blob/feature/updcice_cap/tests/RegressionTests_cheyenne.intel.log) and [GNU](https://github.com/DeniseWorthen/ufs-weather-model/blob/feature/updcice_cap/tests/RegressionTests_cheyenne.gnu.log) for ufs-weather against current develop branch
- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [x] Please provide any additional information or relevant details below:

* Code changes are confined to drivers/nuopc/cmeps with the exceptions of making ``makemask`` public in 
``cicecore/cicedynB/infrastructure/ice_grid.F90`` and minor changes need in ``ice_init`` for CESM
 
* adds drivers/nuopc/cmeps updates from ESCOMP including introduction of mask  file for generating mask when appropriate and changes that allow cice6 to run as an external cesm component
* makes changes in drivers/nuopc/cmeps required for new time manager and introduces a run configurable mesh tolerance error for ufs-weather
* removes forapps/ufs subdirectory which is no longer required by ufs-weather 

These changes are B4B when tested against the current develop branch of ufs-weather-model (00d570e). Tested configurations for ufs-weather-model include multiple configurations, e.g. a  fully active atmosphere (UFSAtm) or a data atmosphere (CDEPS, NEMSDatm) coupled to MOM6 or MOM6+WW3 using the CMEPS mediator. Tests are conducted primarily for different ATM physics options and restart reproducibility of the coupled system. Tested resolutions are C96 ATM 
and 1deg OCN+ICE, C192 ATM and 1/2deg OCN+ICE and C384 ATM and 1/4 deg OCN+ICE. Both Intel and GNU compilers are tested for the coupled system.

### CODE REVIEW

File: cicecore/cicedynB/general/ice_init.F90
Comment (dabail10) [inline cicecore/cicedynB/general/ice_init.F90:453]: Note that this is needed because CESM has restart flags set by driver.

File: cicecore/cicedynB/general/ice_init.F90
Comment (dabail10) [inline cicecore/cicedynB/general/ice_init.F90:1778]: The latlon grid functionality has changed here. We are using a new NUOPC mesh functionality. Are there other groups that use the 'latlon' grid?

File: cicecore/cicedynB/infrastructure/ice_grid.F90
Comment (dabail10) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:43]: makemask is needed in the driver code

File: cicecore/cicedynB/infrastructure/ice_grid.F90
Comment (dabail10) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:383]: Again, is anyone else using latlongrid?

**Review by dabail10 — APPROVED** [review]
There are only two non-driver routines impacted here: ice_init.F90 and ice_grid.F90. I have added comments as a part of the review here. Otherwise, everything else is needed for running the NOAA/EMC UFS app and the CESM.

File: cicecore/cicedynB/general/ice_init.F90
Comment (apcraig) [reply cicecore/cicedynB/general/ice_init.F90:1778]: We need to keep the latlon capability.  F cases use the latlon feature in prescribed ice mode, I assume there may be other users of older version of CESM that need this.  RASM certainly does.

Comment (dabail10) [reply cicecore/cicedynB/general/ice_init.F90:1778]: I didn't realize that RASM still did F compsets! We can certainly keep this. I can make this modification.

**Review by apcraig — CHANGES_REQUESTED** [review]
I think this can be merged after the lonlat mode is recovered.

**Comment by dabail10** [comment]
@DeniseWorthen can you do this? I'm not sure I can modify your pull request.

**Comment by phil-blain** [comment]
> I think this can be merged after the lonlat mode is recovered.

It would be nice however if the merge message would be tweaked to not include the concatenated messages from the 131(!) commits in this PR (which is the default for squash merges)...

**Review by eclare108213 — APPROVED** [review]
I'm fine with the changes to ice_init.F90 and ice_grid.F90.  Someone else should review the nuopc mods.

**Comment by DeniseWorthen** [comment]
@apcraig With further testing, I think I'm seeing the same behaviour as reported in issue [589](https://github.com/CICE-Consortium/CICE/issues/589#issuecomment-833524831). Our use case is with the following:

```
    histfreq       = 'm','d','h','x','x'
    histfreq_n     =  0 , 0 , 6 , 1 , 1
    hist_avg       = .true.
```

However, I am getting hourly history files instead of the expected 6-hourly files. It seems ``elapsed_hours`` is 0 (until the start of a new day) so that ``mod(elapsed_hours,histfreq_n(ns))`` is always 0, and ``write_history`` is true for each new_hour.

**Comment by apcraig** [comment]
@DeniseWorthen, sorry about the hourly output issue.  This is something I could/should have fixed by now, but I thought we'd quickly settle questions about relative vs absolute frequency.  I will fix this soon.  If this is critical for you, let me know and I'll make it even higher priority.  Thanks!

**Review by apcraig — APPROVED** [review]

**Comment by apcraig** [comment]
Just a final checkin @DeniseWorthen, is this ready to merge from your side of things?

**Comment by DeniseWorthen** [comment]
@apcraig We would need to get it fixed prior to the next evaluation we're preparing for the coupled model, but it will be a few weeks before we're ready to run the entire set. Until then, we can make do w/ hourly output. I think it is ready to merge. Thanks.

**Pull request merged**
